### Following tutorial:
https://www.youtube.com/playlist?list=PLSE7WKf_qqo1T5VV1nqXTj2iNiSpFk72T

In [2]:
import sympy as sp

In [5]:
sp.sqrt(2)

sqrt(2)

In [6]:
sp.sqrt(2) ** 2

2

In [9]:
# Make a symbol
x = sp.Symbol("x")
x

x

In [11]:
# A simple expression
2*x + 5

2*x + 5

In [12]:
# The variable name and symbol name do not need to conicide
y = sp.Symbol("z")
y

z

### Simplification

In [13]:
# Sometimes SymPy expressions are automatically simplified
2*x + x - 5

3*x - 5

In [14]:
# Sine and cosine
sp.sin(x)**2 + sp.cos(x)**2

sin(x)**2 + cos(x)**2

In [15]:
# Division
2*x / 6

x/3

In [16]:
# Cancellation in division
x/x

1

In [18]:
# When is the multiplication not carried out?
# SymPy will not expand out parentheses without you explicitly asking for it
expr = x*(x+2)
expr

x*(x + 2)

### Expand expression
To expand an expression, use the expand method

In [19]:
# Expanding the expression
expr.expand()

x**2 + 2*x

In [20]:
# The expression itself does not change
expr

x*(x + 2)

In [21]:
# Can also use expand as a function
sp.expand(expr)

x**2 + 2*x

In [22]:
# Can define multiple symbols at once
sp.symbols("s t")

(s, t)

In [23]:
 # Can check the type
type(sp.symbols("s t"))

tuple

In [24]:
# Can use tuple unpacking to collect the symbols
s, t = sp.symbols("s, t")

In [25]:
poly = t*(s+2)*(t-3)
poly

t*(s + 2)*(t - 3)

In [26]:
# Expanding the curve
poly.expand()

s*t**2 - 3*s*t + 2*t**2 - 6*t

### Factor Expressions
To factor an expression, we will use the factor method

In [27]:
expr = x**2 + 2*x - 15
expr

x**2 + 2*x - 15

In [28]:
# Factoring the polynomial
# Will not work for quintic or higher functions
expr.factor()

(x - 3)*(x + 5)

In [29]:
# Can make many symbols as follows
x_v = sp.symbols("x0:3")

In [30]:
# This is a tuple of symbols
x_v

(x0, x1, x2)

In [31]:
ex = x_v[0]**2 * x_v[1] *x_v[2] + x_v[0] * x_v[1] * x_v[2]
ex

x0**2*x1*x2 + x0*x1*x2

In [33]:
# Factoring the expression
sp.factor(ex)

x0*x1*x2*(x0 + 1)

## Why Data Types in SymPy Matter
[video](https://www.youtube.com/watch?v=SQl-XGFOpFU&list=PLSE7WKf_qqo1T5VV1nqXTj2iNiSpFk72T&index=3)

In [34]:
# Defining a symbol x
x = sp.Symbol('x')

In [36]:
expr = x**(1/3)
expr

x**0.333333333333333

- Firstly, the expression 1/3 is evaluated to be the floating-point number 0.3333...
- Secondly, the SymPy symbol x is raised to the power 0.3333... to form $x^{0.3333\ldots}$

Involving numerics into our symbolic mathematics makes it difficult to recognize terms:

In [37]:
# This should be exactly x^(1/15)
expr ** (1/5)

x**0.0666666666666667

To avoid this, we will introduce the SymPy data types `sp.Integer` and `sp.Rational`

### SymPy's Integer Type

In [39]:
# SymPy's integer type
expr = x ** (sp.Integer(1) / sp.Integer(3))
expr

x**(1/3)

In [40]:
# This should be exactly x^(1/15)
expr ** (sp.Integer(1) / sp.Integer(5))

x**(1/15)

In [42]:
# Even more simple
# If there's an operation between a SymPy object and a python object, 
# the result will always be converted into a SymPy object
expr ** (sp.Integer(1) / 5)

x**(1/15)

In [43]:
# This automatically becomes a SymPy rational object
type(sp.Integer(1) / 5)

sympy.core.numbers.Rational

### SymPy's Rational Type

For specifying rational numbers we can also directly use the `sp.Rational` constructor:

In [45]:
# SymPy's rational object
rational = sp.Rational(1,3)
rational

1/3

Hence we could also have written the following:

In [47]:
expr = (x ** sp.Rational(1,3)) ** sp.Rational(1,5)
expr

x**(1/15)

### Important Constants

Certain constants in mathematics are used over and over again. These are very convenient to have easily accessible in SymPy:

In [49]:
# The constant pi
3*sp.pi

3*pi

In [51]:
# The constant e
sp.E + sp.pi

E + pi

In [52]:
# Infinity
sp.oo

oo

The infinity constant works the way you think it does:

In [53]:
sp.oo - sp.oo

nan

In [54]:
# The imaginary unit
sp.I

I

In [55]:
# Complex arithmetic is implemented
sp.I ** 2

-1

### Common Functions

SymPy has loads of cool functions implemented. Let me go over a few of the most common ones:

In [56]:
# Exponential function
sp.exp(x)

exp(x)

In [57]:
# Eulers identity
sp.exp(sp.pi * sp.I)

-1

In [58]:
# Logarithm function
sp.log(x)

log(x)

In [59]:
# Checking that it is in fact the natural logarithm
sp.exp(sp.log(x))

x

In [60]:
# trigonometric functions
sp.sin(x) + sp.cos(x) + sp.tan(x)

sin(x) + cos(x) + tan(x)

In [61]:
expr = x ** 2 + sp.sin(sp.pi * x) + sp.exp(sp.cos(x))
expr

x**2 + exp(cos(x)) + sin(pi*x)

## Solving Equations

In [this](https://www.youtube.com/watch?v=QJCp1YMDdZ0&list=PLSE7WKf_qqo1T5VV1nqXTj2iNiSpFk72T&index=4) lecture, we are going to learn how to solve equations in SymPy.

**Functions and attributes in this lecture:**
- `sp.Eq()` - Defines a SymPy equation
- `sp.solveset()` - Solves equations.
- `sp.linsolve()` - Solves linear equations.

In [27]:
# Defining symbols x, y, and z
x, y, z = sp.symbols("x, y, z")

## Defining Equations

In mathematics we often write $x^2 = 5$ to define equations. However, in Python the symbol `=` is reserved to be a variable assignment. Double equals `==` is reserved for Boolean comparison.

To define an equation in SymPy we do the following:

In [28]:
# Defines an equation
eq = sp.Eq(x**2, 5)
eq

Eq(x**2, 5)

In [29]:
# Checking the type ofthe equation
type(eq)

sympy.core.relational.Equality

In SymPy we can solve equations with the `solveset()` function:

In [30]:
# Use the solveset method
sp.solveset(eq)

{-sqrt(5), sqrt(5)}

In [31]:
# Checking the type of solveset
type(sp.solveset(eq))

sympy.sets.sets.FiniteSet

In [32]:
# Extracting the first solution
ans = list(sp.solveset(eq,x))
#ans
ans[0]

sqrt(5)

Alternatively, rewrite $x^2 = 5$ as $x^2 - 5 = 0$ and write:

In [33]:
# The same question reformulated
sp.solveset(x**2-5, x)

{-sqrt(5), sqrt(5)}

### Second Example

In [34]:
# Defining the equation
eq2 = sp.Eq(sp.cos(x)-sp.sin(x), 0)
eq2

Eq(-sin(x) + cos(x), 0)

In [35]:
# Solving the equation
sp.solveset(eq2)

Union(ImageSet(Lambda(_n, 2*_n*pi + 5*pi/4), Integers), ImageSet(Lambda(_n, 2*_n*pi + pi/4), Integers))

In [36]:
# Checking out the type of the equation
type(sp.solveset(eq2))

sympy.sets.sets.Union

### Third Example

We know that some equations are only possible to solve numerically. Hence SymPy can't solve them.

In [37]:
# Defining a non-solvable equatoin
eq3 = sp.Eq(sp.cos(x), x)
eq3

Eq(cos(x), x)

In [38]:
# Trying to solve the equation
sp.solveset(eq3)

ConditionSet(x, Eq(-x + cos(x), 0), Complexes)

In [39]:
# Checking the type of the equation
type(sp.solveset(eq3))

sympy.sets.conditionset.ConditionSet

Need to find the solution numerically!

## Solving Systems of Linear Equatoins

In SymPy we can also solve systems of linear equations with the linsolve() function:

In [40]:
# Defining the first equation
eq_lin1 = sp.Eq(x, 3*y + z)
eq_lin1

Eq(x, 3*y + z)

In [41]:
# Defining the second equation
eq_lin2 = sp.Eq(5*x - 3*z, y)
eq_lin2

Eq(5*x - 3*z, y)

In [43]:
# Solving the system of linear equations
sp.linsolve([eq_lin1, eq_lin2], x, y, z)

{(4*z/7, -z/7, z)}

# Simplifying Expressions

In [this](https://www.youtube.com/watch?v=MsP9uEcgLmY&list=PLSE7WKf_qqo1T5VV1nqXTj2iNiSpFk72T&index=5) lecture, we are going to look at simplifying expressions in SymPy. There are several ways of doing this, and we will look at some of the most common ones.

**Functions and attributes in this lecture:**
- `sp.factor()` - Factors expressions
- `sp.expand()` - Expands expressions
- `sp.cancel()` - Puts rational functions in standard canonical form
- `sp.apart()` - Performs a partial fraction decomposition for a function
- `sp.simplify()` - A general simplification function

In [45]:
# Imports
import sympy as sp
# Define the symbols x and y
x, y = sp.symbols('x y')

## Factor and Expand

### Repetition

In [50]:
# Creates a polynomial
poly = x**2 + 5*x + 6
poly

x**2 + 5*x + 6

In [52]:
# Can factor polynomials
poly = poly.factor()
poly

(x + 2)*(x + 3)

In [53]:
# Can expand them again
poly.expand()

x**2 + 5*x + 6

### Expanding Other Expressions

We can use the `expand()` method on other expressions than polynomials

In [54]:
# Expand exponential terms
sp.exp(x + y).expand()

exp(x)*exp(y)

In [55]:
# Trigonometric expression
trig = (sp.cos(x) + sp.sin(y))**2 - sp.cos(x)**2
trig

(sin(y) + cos(x))**2 - cos(x)**2

In [56]:
# Expand to simplify terms
trig.expand()

sin(y)**2 + 2*sin(y)*cos(x)

### Avoiding certain simplifications

Sometimes, we only wish to simplify certain parts of an expression. This can be done with optional arguments in the `expand()` method

In [57]:
# Trigonometric and exponential expression
trig_and_exp = sp.exp(x + y) + (sp.cos(x) + sp.sin(y))**2 - sp.cos(x)**2
trig_and_exp

(sin(y) + cos(x))**2 + exp(x + y) - cos(x)**2

In [58]:
# Expanding everything
trig_and_exp.expand()

exp(x)*exp(y) + sin(y)**2 + 2*sin(y)*cos(x)

If we only want to factor trigonometric terms, then we can write:

In [59]:
# Expanding only trigonometric functions
trig_and_exp.expand(power_exp=False)

exp(x + y) + sin(y)**2 + 2*sin(y)*cos(x)

[Options for the expand method](https://docs.sympy.org/latest/modules/core.html?highlight=expand#sympy.core.expr.Expr.expand)

## Cancel and Apart

The methods `cancel()` and `apart()` are both for simplifying **rational functions** - fractions with polynomials in the numerator and denominator.

### The Cancel Method

The cancel method will take any rational function on the form $\frac{p(x)}{q(x)}$ and factor it so that the polynomials $p(x)$ and $q(x)$ have no common factors

In [60]:
# Defining the polynomials
p = x**3 + 10 * x**2 + 31 * x + 30
q = x**2 + 12 * x + 35
fraction = p / q
fraction

(x**3 + 10*x**2 + 31*x + 30)/(x**2 + 12*x + 35)

In [63]:
# Factors the fraction into the canonical form
fraction.cancel()
# sp.cancel(fraction)

(x**2 + 5*x + 6)/(x + 7)

### Apart Method

The `apart()` method will perform the **partial fraction decomposition** on a rational function. This is useful when integrating a function

In [65]:
# Defining the polynomials
p = x**3 + 10 * x**2 + 31 * x + 30
q = x**2 + 12 * x + 35
fraction = p / q
fraction

(x**3 + 10*x**2 + 31*x + 30)/(x**2 + 12*x + 35)

In [66]:
# Does a partial fraction decomposition of the rational fraction
fraction.apart()

x - 2 + 20/(x + 7)

### The General Simplify

The `simplify()` method is the general purpose simplification method in SymPy. It is the most versatile, but also the slowest and most unspecific.

In [67]:
# Define a massive expression
expr = sp.exp(x+y) + (sp.cos(x) + sp.sin(x))**2 - sp.cos(x) - sp.exp(x)
expr

(sin(x) + cos(x))**2 - exp(x) + exp(x + y) - cos(x)

In [68]:
# General simplification
expr.simplify()

-exp(x) + exp(x + y) + sin(2*x) - cos(x) + 1

If you just want a quick simplification, then try `simplify()`. If you need more fine tuning, then use a more specialized method

# Evaluating Expressions
In [this](https://www.youtube.com/watch?v=i3EoTDYb3WY&list=PLSE7WKf_qqo1T5VV1nqXTj2iNiSpFk72T&index=6) lecture, we are going to learn how to evaluate expressions. We will also take a quick look at specifying that symbols are real, positive, or integers.

** Functions and attributes in this lecture:**
- `.subs()` - Substitute into an expression
- `sp.N()` - Makes a symbolic number numeric